Blind attempt at fixing #572.

jacereda opened the pull request:
Maybe the problem is it can't write to the home directory?

In any case, shouldn't it write to XdgDirectory instead of home?

hellerve commented:
I don’t really know about `XdgDirectory` vs `home`, but I’m seeing that this changes the behavior on OS X from writing to `~/.carp/libs/` to `~/.cache/carp/libs`. I’m personally okay with that change.

We shouldn’t, however, just change the behavior of where we look for `profile.carp` without at least a notice; otherwise I for instance probably wouldn’t figure out where to put the file quickly. (Although of course now I looked)

jacereda commented:
`profile.carp` should probably be loaded from the `XdgConfig` directory.  That would map to `$HOME/.config/carp" on Unix. 

Most software is transitioning from `$HOME/.foo` to `$HOME/.config/foo` nowadays, maybe we should too.

And yes, it should be notified somehow (or provide a backward-compatible fallback for some period).

jacereda commented:
So, what do you think? Should we go all-in and rename the directories, provide backwards-compatibility or forget about it completely?

hellerve commented:
I think we should adhere to best practices, so I guess that yes, we should move from the home directory to the XDG directories! Thanks for your work on this!

hellerve commented on app/Main.hs:
```diff
@@ -80,8 +80,7 @@ main = do setLocaleEncoding utf8
           context <- loadFiles startingContext coreModulesToLoad
-          home <- getHomeDirectory
-          let carpProfile = home ++ "/.carp/profile.carp"
+          carpProfile <- getXdgDirectory XdgConfig "carp/profile.carp"
```
Should we use `</>` here as well?

hellerve commented on app/Main.hs:
```diff
@@ -25,7 +25,7 @@ defaultProject =
           , projectAlreadyLoaded = []
           , projectEchoC = False
-          , projectLibDir = ".carp/libs/"
+          , projectLibDir = "carp/libs/"
```
And here (plus the coming lines)?

jacereda replied:
In this case I think it would be better to stick to unix paths and do a `toNative` where needed, since we don't want conditional compilation for something like configuration.

Or we could just use `System.FilePath.Posix` paths everywhere and perform the `toNative` before calling the OS. 

I'll try to evaluate which one looks better.

hellerve commented:
While we fixed it by #579 I think this is still a good PR to merge!

eriksvedang commented:
I think this broke repl history, no? I'm fine with merge:ing but we need to update the docs also (ideally in the same PR).

hellerve commented:
Yeah, we moved directories here, so both `profile.carp` and the history would have to move.

eriksvedang commented:
The profile file works though.

hellerve commented:
Do you mean the file works without this PR or having it in the old location still works after this PR? The latter would mean that either I’m misremembering or there is something wrong.

eriksvedang commented:
The profile file works with the new location, if I move it there. But the profile history doesn't seem to save at all when using this PR. Does it save for you?

hellerve commented:
I don’t have that problem. In fact it seems that the history file is untouched by this PR (and should be changed). FYI @jacereda the offending line is [this](https://github.com/carp-lang/Carp/blob/master/src/Repl.hs#L75).

It seems like this is an unrelated problem in your setup then, @eriksvedang?

eriksvedang commented:
I guess so, then :)

eriksvedang commented:
I figured it out – for history to work you need a `.carp` in your home directory. I removed this to test out the new setup and then the history stopped working. So this PR should update the history part also, to avoid forcing the user to have two directories, right?

hellerve commented:
Yes, I think we should migrate everything into one directory, and ideally we’d also stop cluttering the home directory. So I think moving everything to XDG is the best way forward!

jacereda commented:
> I figured it out – for history to work you need a `.carp` in your home directory. I removed this to test out the new setup and then the history stopped working. So this PR should update the history part also, to avoid forcing the user to have two directories, right?

I changed it but I'm not sure if the history should be a config file or a cache file.

jacereda commented:
I think I would just use System.FilePath.Posix everywhere (so all paths are Unix paths internally) and only before entering the syscall level perform a `toNative` conversion. Otherwise it's hard not to mix different path separators. 

This would mean creating something like a `Path` module with all the required functionality and centralising there all the filesystem accessing functions. Are you OK with this change?

eriksvedang commented:
> This would mean creating something like a Path module with all the required functionality and centralising there all the filesystem accessing functions. Are you OK with this change?

@jacereda Sounds perfect!

eriksvedang commented:
> I changed it but I'm not sure if the history should be a config file or a cache file.

I'm not sure what effect it has if it's either, but a cache file sounds right. It shouldn't be very important information after all (you wouldn't want to add it to version control for example).

bjorn3 commented:
I think it should be a config file, as caches are normally for rebuildable things, which a history file is not.

jacereda commented:
Changes included as part of https://github.com/carp-lang/Carp/pull/587
